sklearn/metrics/_pairwise_distances_reduction/_radius_neighbors_classmode.pyx.tp

```diff
@@ -0,0 +1,223 @@
+    """
+    cdef:
+        const intp_t[:] class_membership
+        const intp_t[:] unique_labels
+        intp_t outlier_index
+        bint outlier_label_exists
+        bint outliers_exist
+        unsigned char[::1] outliers
+        object outlier_label
+        float64_t[:, :] class_scores
+        WeightingStrategy weight_type
+
+    @classmethod
+    def compute(
+        cls,
+        X,
+        Y,
+        float64_t radius,
+        weights,
+        class_membership,
+        unique_labels,
+        outlier_label=None,
+        str metric="euclidean",
+        chunk_size=None,
+        dict metric_kwargs=None,
+        str strategy=None,
+    ):
+        # Use a generic implementation that handles most scipy
+        # metrics by computing the distances between 2 vectors at a time.
+        pda = RadiusNeighborsClassMode{{name_suffix}}(
+            datasets_pair=DatasetsPair{{name_suffix}}.get_for(X, Y, metric, metric_kwargs),
+            radius=radius,
+            chunk_size=chunk_size,
+            strategy=strategy,
+            weights=weights,
+            class_membership=class_membership,
+            unique_labels=unique_labels,
+            outlier_label=outlier_label,
+        )
+
+        # Limit the number of threads in second level of nested parallelism for BLAS
+        # to avoid threads over-subscription (in GEMM for instance).
+        with threadpool_limits(limits=1, user_api="blas"):
+            if pda.execute_in_parallel_on_Y:
+                pda._parallel_on_Y()
+            else:
+                pda._parallel_on_X()
+
+        return pda._finalize_results()
+
+    def __init__(
+        self,
+        DatasetsPair{{name_suffix}} datasets_pair,
+        const intp_t[:] class_membership,
+        const intp_t[:] unique_labels,
+        float64_t radius,
+        chunk_size=None,
+        strategy=None,
+        weights=None,
+        outlier_label=None,
+    ):
+        super().__init__(
+            datasets_pair=datasets_pair,
+            chunk_size=chunk_size,
+            strategy=strategy,
+            radius=radius,
+        )
+
+        if weights == "uniform":
+            self.weight_type = WeightingStrategy.uniform
+        elif weights == "distance":
+            self.weight_type = WeightingStrategy.distance
+        else:
+            self.weight_type = WeightingStrategy.callable
+
+        self.class_membership = class_membership
+        self.unique_labels = unique_labels
+        self.outlier_index = -1
+        self.outlier_label_exists = False
+        self.outliers_exist = False
+        self.outlier_label = outlier_label
+        self.outliers = np.zeros(self.n_samples_X, dtype=np.bool_)
+
+        cdef intp_t idx
+        if outlier_label is not None:
+            self.outlier_label_exists = True
+            for idx in range(self.unique_labels.shape[0]):
+                if self.unique_labels[idx] == outlier_label:
+                    self.outlier_index = idx
```

Comment:
Is it not possible to use an `outlier_label` value that is not included in the list of unique labels? Mainly, the unique labels are inferred from the provided labels at the call-site, so if a user wanted to mark all outliers with an out-of-distribution label (e.g. `-1`) then they'd need to know a-priori to add it to `unique_labels`.

Either we allow them to provide it independent of `unique_labels`, or clearly document that it **must** be present in `unique_labels`. If we want to keep identical semantics to `RadiusNeighborsClassifier`, we should do the latter. 

From a practitioner perspective, I think this is less than ideal. I imagine, in practice, it is helpful to be able to do this ad-hoc rather than a-priori. That is to say, if you have `unique_labels=[1, 2, 3]` then allow `outlier_label=5` without issue. Even so, that would be a different PR I suppose, since we don't intend for this to really be publicly used...